x-pack/solutions/security/plugins/security_solution/public/asset_inventory/hooks/use_dynamic_entity_flyout.ts

```diff
@@ -0,0 +1,109 @@
+import { useOnExpandableFlyoutClose } from '../../flyout/shared/hooks/use_on_expandable_flyout_close';
+
+interface InventoryFlyoutProps {
+  entity: UniversalEntityEcs;
+  scopeId?: string;
+  contextId?: string;
+}
+
+interface SecurityFlyoutPanelsCommonParams {
+  scopeId?: string;
+  contextId?: string;
+  [key: string]: unknown;
+}
+
+type FlyoutParams =
+  | {
+      id: typeof UniversalEntityPanelKey;
+      params: { entity: InventoryFlyoutProps['entity'] };
+    }
+  | { id: typeof UserPanelKey; params: { userName: string } & SecurityFlyoutPanelsCommonParams }
+  | { id: typeof HostPanelKey; params: { hostName: string } & SecurityFlyoutPanelsCommonParams }
+  | {
+      id: typeof ServicePanelKey;
+      params: { serviceName: string } & SecurityFlyoutPanelsCommonParams;
+    };
+
+const getFlyoutParamsByEntity = (
+  entity: InventoryFlyoutProps['entity'],
+  scopeId?: string,
+  contextId?: string
+): FlyoutParams => {
+  const entitiesFlyoutParams: Record<InventoryFlyoutProps['entity']['type'], FlyoutParams> = {
+    universal: { id: UniversalEntityPanelKey, params: { entity } },
+    user: { id: UserPanelKey, params: { userName: entity.name, scopeId, contextId } },
+    host: { id: HostPanelKey, params: { hostName: entity.name, scopeId, contextId } },
+    service: { id: ServicePanelKey, params: { serviceName: entity.name, scopeId, contextId } },
+  } as const;
+
+  return entitiesFlyoutParams[entity.type];
+};
+
+export const useDynamicEntityFlyout = ({ onFlyoutClose }: { onFlyoutClose: () => void }) => {
+  const { openFlyout, closeFlyout } = useExpandableFlyoutApi();
+  const { notifications } = useKibana().services;
+  useOnExpandableFlyoutClose({ callback: onFlyoutClose });
+
+  const openDynamicFlyout = ({ entity, scopeId, contextId }: InventoryFlyoutProps) => {
+    const entityFlyoutParams = getFlyoutParamsByEntity(entity, scopeId, contextId);
+
+    // User, Host, and Service entity flyouts rely on entity name to fetch required data
+    if (entity.type !== 'universal' && !entity.name) {
+      notifications.toasts.addDanger({
+        title: i18n.translate(
+          'xpack.securitySolution.assetInventory.openFlyout.missingEntityNameTitle',
+          { defaultMessage: 'Missing Entity Name' }
+        ),
+        text: i18n.translate(
+          'xpack.securitySolution.assetInventory.openFlyout.missingEntityNameText',
+          { defaultMessage: 'Entity name is required for User, Host, and Service entities' }
+        ),
+      });
+
+      uiMetricService.trackUiMetric(METRIC_TYPE.CLICK, ASSET_INVENTORY_EXPAND_FLYOUT_ERROR);
+      onFlyoutClose();
+      return;
+    }
+
+    uiMetricService.trackUiMetric(METRIC_TYPE.CLICK, ASSET_INVENTORY_EXPAND_FLYOUT_SUCCESS);
+
+    openFlyout({
+      right: {
+        id: entityFlyoutParams.id || UniversalEntityPanelKey,
+        params: entityFlyoutParams.params,
+      },
+    });
+  };
+
+  const closeDynamicFlyout = () => {
+    closeFlyout();
+  };
```

Comment:
Once we'll add the details and preview panels, this function will expand. right now its mostly for comfort so you dont have to import both this hook and the expandable flyout api separately just to close the flyout